bookmarks/vanilla/src/modules/Bookmark.ts

```diff
@@ -156,90 +164,81 @@ export class Bookmark {
-            const things = getThingAll(updatedDataset)
-
-            return things.map(thing => {
-                return {
-                    url: thing.url,
-                    title: getLiteral(thing, DCTERMS.title)?.value,
-                    link: getLiteral(thing, BOOKMARK.recalls)?.value
-                }
-            }) as IBookmark[]
+            const { title, link, creator, topic } = payload
+
+            thing = setStringNoLocale(thing, DCTERMS.title, title)
+            thing = setStringNoLocale(thing, BOOKMARK.recalls, link)
+            if (creator) thing = setNamedNode(thing, DCTERMS.creator, namedNode(creator))
+            if (topic) thing = setNamedNode(thing, BOOKMARK.hasTopic, namedNode(topic))
+            thing = setStringNoLocale(thing, "http://purl.org/dc/terms/updated", new Date().toISOString())
+
+            thing = setUrl(thing, RDF.type, BOOKMARK.Bookmark)
+
+            const updatedBookmarkList = setThing(ds, thing);
+            await saveSolidDatasetAt(indexUrl, updatedBookmarkList, { fetch: session.fetch });
+
+            return { url, ...payload }
         }
 
     };
 
 
-    private static mapBookmark(thing: ThingPersisted): Bookmark {
-        const obj: IBookmark = {
-            url: thing.url,
-            title: this.mapTitle(thing),
-            link: this.mapLink(thing)
-        };
-        if (this.mapCreated(thing)) {
-            obj.created = this.mapCreated(thing);
-        }
-        if (this.mapUpdated(thing)) {
-            obj.updated = this.mapUpdated(thing);
-        }
-        if (this.mapCreator(thing)) {
-            obj.creator = this.mapCreator(thing);
-        }
-        if (this.mapTopic(thing)) {
-            obj.topic = this.mapTopic(thing);
+    private static mapBookmark(thing: ThingPersisted): IBookmark {
+        const url = thing.url
+        const title = this.mapTitle(thing)
+        const link = this.mapLink(thing)
+        const topic = this.mapTopic(thing)
+        const created = this.mapCreated(thing)
+        const updated = this.mapUpdated(thing)
+        const creator = this.mapCreator(thing)
+
+        return {
+            url,
+            title,
+            link,
+            ...(topic && { topic }),
+            ...(created && { created }),
+            ...(updated && { updated }),
+            ...(creator && { creator }),
         }
-        return obj;
     }
     private static mapTitle(thing: ThingPersisted): string {
         return (
             getLiteral(thing, DCTERMS.title)?.value ??
-            getLiteral(thing, RDFS.label)?.value ??
-            ""
+            getLiteral(thing, RDFS.label)?.value ?? ""
         );
     }
     private static mapLink(thing: ThingPersisted): string {
         return (
             getLiteral(thing, BOOKMARK.recalls)?.value ??
-            getNamedNode(thing, BOOKMARK.recalls)?.value ??
-            ""
+            getNamedNode(thing, BOOKMARK.recalls)?.value ?? ""
```

Comment:
An empty string is not a valid URL. I think we have throw an exception in this case. A bookmark without a URL is not a bookmark.

Reply:
sure, done